cdap-proto/src/main/java/co/cask/cdap/proto/Id.java

```diff
@@ -59,33 +53,19 @@ private static boolean isValidDatasetId(String datasetId) {
     return datasetIdCharMatcher.matchesAllOf(datasetId);
   }
 
-  public String getIdType() {
-    return getType(this.getClass());
+  public static String getIdType(Class<? extends Id> cls) {
+    return cls.getSimpleName().toLowerCase();
   }
 
-  protected String getIdForRep() {
-    return getId();
-  }
-
-  public String getIdRep() {
-    Id parent = getParent();
-    if (parent == null) {
-      return getIdType() + ":" + getIdForRep();
-    } else {
-      return parent.getIdRep() + "/" + getIdType() + ":" + getIdForRep();
-    }
+  public static boolean isType(String string, Class<? extends Id> expectedType) {
+    String[] parts = string.split(":", 2);
+    return parts.length == 2 && getIdType(expectedType).equals(parts[0]);
   }
 
-  @Override
-  public String toString() {
-    return getIdRep();
+  public String getIdType() {
+    return getIdType(getClass());
   }
 
-  @Nullable
-  protected abstract Id getParent();
-
-  public abstract String getId();
```

Comment:
getId is a getter and should simply return the value of the id field. Actually it would be better to rename id to name and then you'd have getName() return "flow1", getFullName() return "namespace1.app1.flow.flow1", and toString() returns  "program:namespace1.app1.flow.flow1"